File: src/image_gen/webui/static/assets/js/workspace/responsive.js
export const WORKSPACE_WIDTH_CLASSES = Object.freeze(["wide", "standard", "compact", "narrow"]);
export const WORKSPACE_WIDTH_THRESHOLDS = Object.freeze({
  narrowMax: 719,
  compactMax: 1049,
  standardMax: 1450,
});
export const WORKSPACE_REPRESENTATIVE_WIDTHS = Object.freeze({
  wide: 1600,
  standard: 1280,
  compact: 960,
  narrow: 680,
});

function normalizedWidthClass(value) {
  const token = String(value || "").trim().toLowerCase();
  return WORKSPACE_WIDTH_CLASSES.includes(token) ? token : "wide";
}

function normalizedSpan(value, fallback = 12) {
  const span = Number(value);
  if (!Number.isFinite(span)) return fallback;
  return Math.max(1, Math.min(12, Math.round(span)));
}

export function classifyWorkspaceWidth(width) {
  const value = Math.max(0, Number(width) || 0);
  if (value > WORKSPACE_WIDTH_THRESHOLDS.standardMax) return "wide";
  if (value > WORKSPACE_WIDTH_THRESHOLDS.compactMax) return "standard";
  if (value > WORKSPACE_WIDTH_THRESHOLDS.narrowMax) return "compact";
  return "narrow";
}

export function responsiveGridSpan(baseSpan, widthClass, { minGridSpan = 1, maxGridSpan = 12 } = {}) {
  const width = normalizedWidthClass(widthClass);
  const minimum = normalizedSpan(minGridSpan, 1);
  const maximum = Math.max(minimum, normalizedSpan(maxGridSpan, 12));
  const base = Math.max(minimum, Math.min(maximum, normalizedSpan(baseSpan, 12)));
  let requested = base;
  if (width === "standard") requested = base >= 8 ? 12 : Math.max(6, base);
  if (width === "compact") requested = base >= 7 ? 12 : Math.max(6, base);
  if (width === "narrow") requested = 12;
  return Math.max(minimum, Math.min(maximum, requested));
}

export function responsivePresentationSpan(effectiveSpan, widthClass, shellState = "expanded", { minGridSpan = 1, maxGridSpan = 12 } = {}) {
  const width = normalizedWidthClass(widthClass);
  const minimum = normalizedSpan(minGridSpan, 1);
  const maximum = Math.max(minimum, normalizedSpan(maxGridSpan, 12));
  const base = Math.max(minimum, Math.min(maximum, normalizedSpan(effectiveSpan, 12)));
  if (String(shellState || "expanded").trim().toLowerCase() !== "side") return base;
  const target = width === "wide" ? 2 : (width === "standard" ? 3 : (width === "compact" ? 6 : 12));
  return Math.max(minimum, Math.min(maximum, target));
}

export function resolveResponsiveVariant(descriptor, widthClass, preferredVariant = "", placementOverrides = {}) {
  const width = normalizedWidthClass(widthClass);
  const supported = [...(descriptor?.supportedVariants || [])].map(String);
  const preferred = String(preferredVariant || descriptor?.defaultVariant || "standard");
  const responsive = descriptor?.responsive && typeof descriptor.responsive === "object" ? descriptor.responsive : {};
  const overrides = placementOverrides && typeof placementOverrides === "object" ? placementOverrides : {};
  const desired = width === "wide"
    ? preferred
    : String(overrides[width] || responsive[width] || preferred || "standard");
  if (supported.includes(desired)) return { variant: desired, resolution: "preferred" };
  if (supported.includes("standard")) return { variant: "standard", resolution: "standard_fallback" };
  if (supported.includes(preferred)) return { variant: preferred, resolution: "preferred_fallback" };
  const defaultVariant = String(descriptor?.defaultVariant || "standard");
  if (supported.includes(defaultVariant)) return { variant: defaultVariant, resolution: "default_fallback" };
  return { variant: supported[0] || desired || "standard", resolution: "first_supported_fallback" };
}

export function responsiveWorkspacePlacements(layout, descriptorLookup, widthClass) {
  const source = Array.isArray(layout?.components) ? layout.components : [];
  return source.map((placement, order) => {
    const descriptor = typeof descriptorLookup === "function" ? descriptorLookup(placement.componentId) : null;
    if (!descriptor) return null;
    const baseSpan = normalizedSpan(placement.span ?? placement.position?.columnSpan, descriptor.defaultGridSpan || 12);
    const resolved = resolveResponsiveVariant(descriptor, widthClass, placement.variant, placement.responsive);
    const responsiveSpan = responsiveGridSpan(baseSpan, widthClass, descriptor);
    const effectiveSpan = responsivePresentationSpan(responsiveSpan, widthClass, placement.shellState, descriptor);
    return {
      ...placement,
      order: Number.isFinite(Number(placement.order)) ? Number(placement.order) : order,
      baseSpan,
      responsiveSpan,
      effectiveSpan,
      baseVariant: placement.variant || descriptor.defaultVariant,
      effectiveVariant: resolved.variant,
      variantResolution: resolved.resolution,
      widthClass: normalizedWidthClass(widthClass),
    };
  }).filter(Boolean);
}

function measuredWidth(root) {
  const rect = root?.getBoundingClientRect?.();
  return Math.max(0, Number(rect?.width || root?.clientWidth || 0));
}

export function bindWorkspaceWidthObserver(root, { onChange = null } = {}) {
  if (!root) return null;
  let lastClass = "";
  let lastWidth = -1;
  const apply = (width = measuredWidth(root)) => {
    const rounded = Math.max(0, Math.round(Number(width) || 0));
    const widthClass = classifyWorkspaceWidth(rounded);
    root.dataset.workspaceWidthClass = widthClass;
    root.dataset.workspaceWidth = String(rounded);
    const classChanged = widthClass !== lastClass;
    const previousWidthClass = lastClass || null;
    lastWidth = rounded;
    if (classChanged) {
      lastClass = widthClass;
      const detail = { width: rounded, widthClass, previousWidthClass };
      onChange?.(detail);
      root.dispatchEvent(new CustomEvent("workspace-width-class-change", { bubbles: true, detail }));
    }
    return { width: lastWidth, widthClass };
  };

  const observer = typeof ResizeObserver === "function"
    ? new ResizeObserver((entries) => {
        const entry = entries[entries.length - 1];
        const width = entry?.contentRect?.width ?? measuredWidth(root);
        apply(width);
      })
    : null;
  observer?.observe(root);
  apply();
  return Object.freeze({
    refresh: () => apply(),
    disconnect: () => observer?.disconnect(),
  });
}

File: src/image_gen/webui/workspace/responsive.py
from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from typing import Mapping, TYPE_CHECKING

from .models import WorkspaceComponentDescriptor, WorkspaceLayout

if TYPE_CHECKING:
    from .registry import WorkspaceRegistry


class WorkspaceWidthClass(str, Enum):
    WIDE = "wide"
    STANDARD = "standard"
    COMPACT = "compact"
    NARROW = "narrow"


WORKSPACE_WIDTH_CLASSES = tuple(item.value for item in WorkspaceWidthClass)
WORKSPACE_WIDTH_THRESHOLDS = {
    "narrowMax": 719,
    "compactMax": 1049,
    "standardMax": 1450,
}
WORKSPACE_REPRESENTATIVE_WIDTHS = {
    WorkspaceWidthClass.WIDE.value: 1600,
    WorkspaceWidthClass.STANDARD.value: 1280,
    WorkspaceWidthClass.COMPACT.value: 960,
    WorkspaceWidthClass.NARROW.value: 680,
}


def classify_workspace_width(width: float | int) -> WorkspaceWidthClass:
    value = max(0.0, float(width))
    if value > WORKSPACE_WIDTH_THRESHOLDS["standardMax"]:
        return WorkspaceWidthClass.WIDE
    if value > WORKSPACE_WIDTH_THRESHOLDS["compactMax"]:
        return WorkspaceWidthClass.STANDARD
    if value > WORKSPACE_WIDTH_THRESHOLDS["narrowMax"]:
        return WorkspaceWidthClass.COMPACT
    return WorkspaceWidthClass.NARROW


def responsive_grid_span(
    base_span: int,
    width_class: WorkspaceWidthClass | str,
    *,
    minimum: int = 1,
    maximum: int = 12,
) -> int:
    width_value = WorkspaceWidthClass(str(getattr(width_class, "value", width_class)))
    base = max(minimum, min(maximum, int(base_span)))
    if width_value is WorkspaceWidthClass.WIDE:
        requested = base
    elif width_value is WorkspaceWidthClass.STANDARD:
        requested = 12 if base >= 8 else max(6, base)
    elif width_value is WorkspaceWidthClass.COMPACT:
        requested = 12 if base >= 7 else max(6, base)
    else:
        requested = 12
    return max(minimum, min(maximum, requested))


def responsive_presentation_span(
    effective_span: int,
    width_class: WorkspaceWidthClass | str,
    *,
    shell_state: str = "expanded",
    minimum: int = 1,
    maximum: int = 12,
) -> int:
    width_value = WorkspaceWidthClass(str(getattr(width_class, "value", width_class)))
    base = max(minimum, min(maximum, int(effective_span)))
    if str(shell_state or "expanded").strip().lower() != "side":
        return base
    target = {
        WorkspaceWidthClass.WIDE: 2,
        WorkspaceWidthClass.STANDARD: 3,
        WorkspaceWidthClass.COMPACT: 6,
        WorkspaceWidthClass.NARROW: 12,
    }[width_value]
    return max(minimum, min(maximum, target))


def resolve_responsive_variant(
    descriptor: WorkspaceComponentDescriptor,
    width_class: WorkspaceWidthClass | str,
    *,
    preferred_variant: str = "",
    placement_overrides: Mapping[str, object] | None = None,
) -> tuple[str, str]:
    width_value = WorkspaceWidthClass(str(getattr(width_class, "value", width_class))).value
    preferred = str(preferred_variant or descriptor.default_variant or "standard")
    overrides = placement_overrides or {}
    desired = ""
    if width_value != WorkspaceWidthClass.WIDE.value:
        desired = str(overrides.get(width_value) or descriptor.responsive_variants.get(width_value) or "").strip()
    if not desired:
        desired = preferred

    supported = tuple(descriptor.supported_variants)
    if desired in supported:
        return desired, "preferred"
    if "standard" in supported:
        return "standard", "standard_fallback"
    if preferred in supported:
        return preferred, "preferred_fallback"
    if descriptor.default_variant in supported:
        return descriptor.default_variant, "default_fallback"
    return supported[0], "first_supported_fallback"


@dataclass(frozen=True)
class WorkspaceResponsivePlacement:
    component_id: str
    instance_id: str
    width_class: str
    base_span: int
    effective_span: int
    base_variant: str
    effective_variant: str
    variant_resolution: str

    def to_dict(self) -> dict[str, object]:
        return {
            "componentId": self.component_id,
            "instanceId": self.instance_id,
            "widthClass": self.width_class,
            "baseSpan": self.base_span,
            "effectiveSpan": self.effective_span,
            "baseVariant": self.base_variant,
            "effectiveVariant": self.effective_variant,
            "variantResolution": self.variant_resolution,
        }


def build_responsive_layout_plan(
    layout: WorkspaceLayout,
    registry: "WorkspaceRegistry",
    width: float | int,
) -> tuple[WorkspaceResponsivePlacement, ...]:
    width_class = classify_workspace_width(width)
    planned: list[WorkspaceResponsivePlacement] = []
    for placement in layout.components:
        descriptor = registry.component(placement.component_id)
        if descriptor is None:
            continue
        variant, resolution = resolve_responsive_variant(
            descriptor,
            width_class,
            preferred_variant=placement.variant,
            placement_overrides=placement.responsive,
        )
        responsive_span = responsive_grid_span(
            placement.position.column_span,
            width_class,
            minimum=descriptor.min_grid_span,
            maximum=descriptor.max_grid_span,
        )
        planned.append(
            WorkspaceResponsivePlacement(
                component_id=placement.component_id,
                instance_id=placement.instance_id,
                width_class=width_class.value,
                base_span=placement.position.column_span,
                effective_span=responsive_presentation_span(
                    responsive_span,
                    width_class,
                    shell_state=placement.shell_state,
                    minimum=descriptor.min_grid_span,
                    maximum=descriptor.max_grid_span,
                ),
                base_variant=placement.variant,
                effective_variant=variant,
                variant_resolution=resolution,
            )
        )
    return tuple(planned)


def workspace_responsive_contract_payload() -> dict[str, object]:
    return {
        "widthClasses": list(WORKSPACE_WIDTH_CLASSES),
        "thresholds": dict(WORKSPACE_WIDTH_THRESHOLDS),
        "representativeWidths": dict(WORKSPACE_REPRESENTATIVE_WIDTHS),
        "measurement": "workspace-container",
        "gridColumns": 12,
        "snapGridPixels": 8,
    }
